Update Native Functions documentation (#252)
grafana/tanka: docs/docs/jsonnet/native-functions.md

# Native Functions

Tanka extends Jsonnet using _native functions_, offering additional functionality not yet available in the standard library.

To use them in your code, you need to access them using `std.native` from the standard library:

```jsonnet
{

>   deployment:  std.native('<name>')('<arguments>'),

Reviewer: ```suggestion
  someField:  std.native('<name>')(<arguments>),
```

Arguments are not always a string

Author: Good catch

}
```

`std.native` takes the native function's name as a `string` argument and returns a `function`, which is called using the second set of parentheses.

## parseJson

### Signature

```ts
parseJSON(string json) Object
```

> `parseJSON` wraps `json.Unmarshal` to convert a json string into a dict.

Reviewer: ```suggestion
`parseJSON` parses a json string and returns the respective Jsonnet type (`Object`, `Array`,  etc).
```

Thinking about this, nobody cares which underlying Go function  is used, so let's describe the behaviour instead

Author: Agreed

### Examples

```jsonnet
{

>   parseJsonArray: std.native('parseJson')('[0, 1, 2]'),
>   parseJsonObject: std.native('parseJson')('{ "foo": "bar" }'),

Reviewer: ```suggestion
  array: std.native('parseJson')('[0, 1, 2]'),
  object: std.native('parseJson')('{ "foo": "bar" }'),
```

}
```

Evaluating with Tanka results in the JSON:

```json
{

>   "parseJsonArray": [0, 1, 2],
>   "parseJsonObject": {

Reviewer: ```suggestion
  "array": [0, 1, 2],
  "object": {
```

# Native Functions

> Native functions are Jsonnet functions implemented in Go rather than Jsonnet itself. Tanka extends the Jsonnet native functions with some common helper utilities.

Reviewer: ```suggestion
Tanka extends Jsonnet using *native functions*, offering additional functionality not yet available in the standard library.
```

> To use a native function in your Jsonnet code, you need to use the `native` function from the Jsonnet standard library. For example, to parse a Kubernetes deployment YAML file into a Jsonnet object:

Reviewer: ```suggestion
To use them in your code, you need to access them using `std.native` from the standard library:
```

```jsonnet
{

>   deployment:  std.native('parseYaml')(importstr 'deployment.yaml'),

Reviewer: ```suggestion
  deployment: std.native('<name>')(<arguments>),
```

}
```

> The argument to `std.native` is the name of the native function to lookup. In this case, it is `parseYaml`. The returned function is then evaluated with the arguments in the second set of parentheses. This is the `deployment.yaml` file imported as a string using the Jsonnet keyword `importstr`.

Reviewer: ```suggestion
`std.native` takes the native function's name as a `string` argument and returns a `function`, which is called using the second set of parentheses.
```

## parseJson

### Usage

```jsonnet

> std.native('parseJson')(json)

Reviewer: I'd prefer to revert back to the function signatures, as they can come handy as reference material, especially regarding argument types.

When replacing the `Usage` section with `Signature`, your `Examples` section should still provide enough information on how to use it!

Author: That's true, I did hesitate when replacing the signature with the usage because losing the typing did make the native functions arguments less clear and the workaround of mentioning the types in the subsequent comment felt a bit clunky. Like in the `regexSubst` comment:
```
regexSubst` replaces all matches of the re2 regular expression `regex` in the `src` string with the `repl` string.
```

I'll make the switch back to `Signature` and revert the associated comments too.

```

`parseJson` parses a JSON string and returns the JSON element.

### Examples

```jsonnet
{
  parseJsonArray: std.native('parseJson')('[0, 1, 2]'),
  parseJsonObject: std.native('parseJson')('{ "foo": "bar" }'),
}
```

Evaluating with Tanka results in the JSON:

```json
{
  "parseJsonArray": [0, 1, 2],
  "parseJsonObject": {
    "foo": "bar"
  }
}
```

## parseYaml

### Usage

```jsonnet
std.native('parseYaml')(yaml)
```

`parseYaml` parses a string of YAML document(s) and returns an array of JSON objects. If `yaml` contains only a single document, a single value array will be returned.

### Examples

```jsonnet
{

>   parseYaml: std.native('parseYaml')(|||

Reviewer: Might be confusing to have the nativeFuncs name as the object key here – which is obviously not a requirement. 

Maybe rename all keys to something describing the content, such as `yaml`, `escaped`, `matched`, etc?

Author: That's true, and your suggestion for the renames will make that a lot clearer and understandable for sure.

a set of dicts. If `yaml` only contains a single document, a single value array

> will be returned.

Reviewer: ```
`parseYAML` converts a string of YAML documents (YAML stream) into native Jsonnet types.
Always returns an `Array`, even if only a single document is passed.
